test/xbgpu/test_engine.py

```diff
@@ -565,11 +540,23 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
             )
 
         with PromDiff(namespace=METRIC_NAMESPACE) as prom_diff:
-            # Generate one extra chunk to simulate an incomplete accumulation
-            # to check that dumps are aligned correctly - even if the first
-            # received batch is from the middle of an accumulation.
-            batch_start_index = first_batch_index * max_heap_acc_threshold - HEAPS_PER_FENGINE_PER_CHUNK
-            batch_end_index = (first_batch_index + min_full_accumulations) * max_heap_acc_threshold
+            # NOTE: The product of `heap_acc_thresholds` is used in two ways
+            # below. Both uses are to ensure there is a whole number of
+            # accumulations for *both* XPipelines. The first usage is
+            # dual-purpose once more:
+            # - In addition to the above, the arbitrary start position for data
+            #   in this test dictates the first received batch to be in the
+            #   middle of an accumulation.
+            # - More accurately, it forces the first accumulation (to be
+            #   processed) to only contain one batch of data. This ensures
+            #   we test that output dumps are aligned correctly, despite
+            #   the first data processed not being on an accumulation
+            #   boundary.
+            heap_acc_thresholds = [corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs]
+            first_batch_index = 12 * np.product(heap_acc_thresholds)
+            batch_start_index = first_batch_index - HEAPS_PER_FENGINE_PER_CHUNK
+            batch_end_index = first_batch_index + np.product(heap_acc_thresholds)
```

Comment:
I'd suggest just a bit of a style tweak, and a comment to clarify the selection of the arbitrary constant (12).

```suggestion
            n_heaps = np.product([corrprod_output.heap_accumulation_threshold for corrprod_output in corrprod_outputs])
            first_batch_index = 12 * n_heaps  # Just somewhere arbitrarily that isn't zero
            batch_start_index = first_batch_index - HEAPS_PER_FENGINE_PER_CHUNK
            batch_end_index = first_batch_index + n_heaps
```